File: MAY 24/08 May/Root to Leaf Paths gfg.cpp
class Solution {
    private:
    void traverse(Node* root, vector<int> &path, vector<vector<int>> &paths)
    {
        if(root==nullptr)
            return;
        
        path.push_back(root->data);
        if(root->left==nullptr && root->right==nullptr)
        {
            paths.push_back(path);
        }
        else
        {
            traverse(root->left, path, paths);    
            traverse(root->right, path, paths);
        }
        path.pop_back();
    }
    public:
    vector<vector<int>> Paths(Node* root) {
        // code here
        vector<vector<int>> paths;
        vector<int> path;
        traverse(root, path, paths);
        return paths;
    }
};

File: MAY 2025/07 May/Root to Leaf Paths gfg.java

class Solution {
    private static void traverse(ArrayList<ArrayList<Integer>> paths, ArrayList<Integer> path, Node root)
    {
        if(root == null)
            return;
        path.add(root.data);  
        
        if(root.left == null && root.right == null)
            paths.add(new ArrayList<>(path));

        
        traverse(paths, path, root.left);
        traverse(paths, path, root.right);
        
        path.remove(path.size()-1);
    }
    
    public static ArrayList<ArrayList<Integer>> Paths(Node root) {
        // code here
        ArrayList<ArrayList<Integer>> paths = new ArrayList<>();
        ArrayList<Integer> path = new ArrayList<>();
        
        traverse(paths, path, root);
        
        return paths;
    }
}
